Improve flow of importing a dataset

philippotto opened the pull request:
Since the import of a dataset is a crucial step to be able to use wk, I invested some time to polish that flow and to make it bulletproof™. This PR:
- cleans up and re-structures the import dialog
- adds help tooltips to most of the fields
- provides a "simple" mode for editing most important data properties (such as scale and layer-specific properties)
- adds more thorough validation
- adds a confirm-modal if the user uploads a dataset without assigning teams to it

General feedback is very appreciated. Maybe @normanrz also wants to give feedback on the wording.

### URL of deployed dev instance (used for testing):
- https://improveimport.webknossos.xyz

### Steps to test:
@fm3  Since this PR is rather big and changes a crucial part of wk, it would be cool if you could  support @daniel-wer with testing this (at least some spot checks).

#### Combinations
There are multiple combinations, in which the import view can be tested: 
- Upload vs. Edit
  - The import view shows after uploading a dataset or after moving a new dataset into the appropriate folder
  - The same view is also reused when editing an already imported dataset. This mode shows an additional tab (default config) and more properties in the "General" tab (display name etc.)
- `Datasource.json`s in imported dataset can be:
  - missing (wk should guess the values then)
  - invalid (wk should guess the values then; the warning will be a bit more "severe")
  - valid (the UI will tell you that the import is already done after upload. the form is only shown for non-mandatory editing purposes)
- Upload via Browser vs. Putting the dataset directly into the right folder
- KNOSSOS vs. WKW: the view should work when handling knossos **and** wkw datasets

#### Functionality (sorted by importance):
- in general, editing properties and submitting the form should import/save the dataset (provided that the values are valid)
- datasource.json-related properties should be modifiable via the "advanced" mode as well as "basic" mode
  - manually changing properties in one mode, should change the properties automatically in the other mode
  - submitting the form without switching between the modes should save the newest values
  - when the JSON string is invalid, switching back to "basic" mode should not be possible
- handling of invalid fields
  - invalid fields (e.g., bounding box with zero width, or zero scale) should be highlighted (in the tab title as well as in the form itself)
  - submitting the form with invalid fields, should switch to the tab which has invalid fields
- no allowed teams
  - when no allowed teams are selected, this should be warned about via an orange label under the select-input as well as an orange icon in the tab title
  - when submitting the form with no allowed teams, a confirm-modal should ask if the user is sure about that


### Issues:
- fixes #2852

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [X] ~Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~
- [X] Ready for review

philippotto commented:
@daniel-wer The import view and the corresponding code became quite a beast. I'm not really satisfied with the code, but I'm also not sure how to reduce its complexity. Feedback is highly appreciated :)

daniel-wer commented:
Things I noticed during testing:

  - [ ] While in the advanced mode I hit my "Tab" key in the Dataset Configuration input, which scrolls the viewport to the far right, without having the possibility to scroll back.
![dataset-1](https://user-images.githubusercontent.com/1702075/42582766-407b6a94-8530-11e8-9bae-2358f437256c.png)
  - [ ] If the boundingBox (or Scale) is invalid (5 instead of 6 values), when blurring, all values are replaced by 0 and the error message is shown. Ideally my entered values would remain so I can adjust/complete them.
  - [ ] The Bounding box field should be wider, so that values like `4480, 4736, 1920, 10000, 10000, 10000` fit in there at once.
  - [ ] The value for segmentation opacity should be enforced to be between 0 and 100
  - [ ] What's the reason that the view configuration and things like the display name cannot be configured when importing a dataset. I think the discoverability of the possibility to adjust these values is not very good (only when editing the dataset afterwards).

daniel-wer commented on app/assets/javascripts/dashboard/dataset/dataset_import_view.js:
```diff
@@ -239,116 +286,120 @@ class DatasetImportView extends React.PureComponent<Props, State> {
+    );
+    const _hasNoAllowedTeams = this.hasNoAllowedTeams();
+    const hasNoAllowedTeamsWarning = _hasNoAllowedTeams ? (
+      <Tooltip title="Please double-check some fields here.">
```
I see now that this is the tooltip that is displayed when hovering over the tab. Please disregard my previous comment :)

daniel-wer commented:
@fm3 Because it may be relevant for this PR, I vaguely remember that datasets were imported automatically (if all jsons were present) at one point in the past. Do you know whether this functionality still exists?

philippotto commented on app/assets/javascripts/dashboard/dataset/dataset_import_view.js:
```diff
@@ -152,41 +127,123 @@ class DatasetImportView extends React.PureComponent<Props, State> {
+    }
+    // Switch to the earliest, problematic tab
+    let problematicTab: ?TabKeyType = null;
+    if (validationSummary.data) {
```
Good idea! `findKey` loses the type information, but I changed it to 

```
const problematicTab = _.find(
      ["data", "general", "defaultConfig"],
      key => validationSummary[key],
    );
```

now, which gives good type checking (at the cost of repeating the name). However, it's definitely more concise than what's it been before.

philippotto commented:
> While in the advanced mode I hit my "Tab" key in the Dataset Configuration input, which scrolls the viewport to the far right, without having the possibility to scroll back.
dataset-1

Oh dear... I guess antd tries to be smart and to switch to the next input element which is in another tab. Forms are probably not designed to be spread across multiple tabs :S I'll have a look tomorrow whether we can mitigate this. Good news: shift + tab will bring you back :laughing: 

> If the boundingBox (or Scale) is invalid (5 instead of 6 values), when blurring, all values are replaced by 0 and the error message is shown. Ideally my entered values would remain so I can adjust/complete them.

Yeah, I also noticed that. The vector input already existed and is probably designed like that so that it will never propagate invalid values (used in wk core). I'll see whether I can integrate a flag à la `allowInvalidState`.

> In the advanced mode, often the Switch (advanced/simple) is still enabled, although the datasource json is not valid. Try to remove one value from the scale, for example, to reproduce this.

It's not a bug, it's a feature :P If the JSON string is invalid in the sense that it cannot be parsed, you cannot switch back to "simple". That way, we avoid issues where we try to parse the JSON to sync the values. If the JSON is invalid in the sense of being non-conform to the format we require, you can still switch back, since you can fix most things in the simple mode, as well. This is more or less necessary, since the initial json is sometimes invalid (missing segment id). In that case, we still want to show the simple UI.

> What's the reason that the view configuration and things like the display name cannot be configured when importing a dataset. I think the discoverability of the possibility to adjust these values is not very good (only when editing the dataset afterwards).

Yeah, I agree with you. I wonder whether this had some special reason at some point. Maybe @normanrz or @fm3 know a reason?

philippotto commented:
I incorporated (almost) all your feedback, @daniel-wer! Thanks for your thoughtful input :)

> If the boundingBox (or Scale) is invalid (5 instead of 6 values), when blurring, all values are replaced by 0 and the error message is shown. Ideally my entered values would remain so I can adjust/complete them.

This is the only thing, I did slightly differently. Since the a lot of code assumes that the InputVector components will always hold valid values, I didn't want to rewrite that, as it would duplicate the validation (within the component **and** outside the component). Instead, I changed the InputVector so that invalid values are not completely thrown away but are reused instead. For example, inputting "1, 2, 3, 4" for a vector3 will yield "1, 2, 3". "1, 2" will yield "1, 2, 0". And non-numbers will just be removed ("1ab2, 3, 4" will become "12, 3, 4"). I think this is a good trade-off.

@fm3 Feel free to test the PR now (I need to update the dev instance, though). I also added some more cases to the test plan :tada:

fm3 commented:
Some things I noticed (not sure if all of those need to be adressed)

- in scale field, type `1,2,`, then leave it, creates `1, 2, NaN`
- typing `asdf` creates `NaN, 0, 0`
- same with boundingboxes
- would be nice if the `must not be zero` warnings would disappear as soon as I type a valid value (not only on blur)

- generally, sometimes the timing of snapping to default/corrected values feels slightly unintuitive. Example (albeit unlikely scenario): In advanced tab, remove one line from scale. I get `instance.scale does not meet minimum length of 3`. Then, switch to simple, fix scale in input field. The exclamation point next to the Data tab header does not vanish (even on blur), yet I can click import and it succeeds.


- If I put an invalid datasource-json on disk, I get `The datasource-properties.json on disk is invalid. The values below are guessed by webKnossos. Please review all properties before importing the dataset.Status: Error: Invalid json format in binaryData/Connectomics department/…` – I would like a line break (or at least a space) before Status

- I got `, 0, 0` in View Configuration Position field, somehow I managed to save (null, 0, 0) there, don’t exactly remember the steps, sorry.

- after upload dataset, I am redirected to import page. after clicking save there, I am redirected back to the upload page. I would have expected the datasets list, but that may be intentional

philippotto commented:
Thanks for the testing and for your feedback, @fm3! I fixed most of the things you mentioned.

> would be nice if the must not be zero warnings would disappear as soon as I type a valid value (not only on blur)

Unfortunately, this is not easy to realize with the current code. I'm not 100% satisfied with it, either, but I'd like to leave it as is for now, as it's still usable.

> after upload dataset, I am redirected to import page. after clicking save there, I am redirected back to the upload page. I would have expected the datasets list, but that may be intentional

I think, it's been like that for quite a while now. We navigate to the last page visited, which is usually the import view. We should probably change this to just redirect to the dashboard dataset page. However, my other PR which integrates the import flow into the onboarding flow changes things in this regard. I'd like to do the change in that PR (or in a new one).

philippotto commented:
@daniel-wer From point of view, the PR should be ready to merge. Do you agree?
